### ISSUE TITLE
Make Bitmap, Image, Metafile, Icon, Font serializable for Winforms use

### ISSUE THREAD

**Issue by ericstj** [dotnet/corefx#30183]
/cc @Tanya-Solyanik @morganbr 

The WinForms app model / designer relies on placing System.Drawing types in resources and serializing those using BinaryFormatter.  As such we need to add back [Serializable] to these types.

Alternatively the WinForms designer (and resources designer, likely) could use a different serialization format for the blobs embedded in the resx.  This would require a coordinated effort across framework / designers / msbuild but could mitigate an increased exposure to problematic binaryformatter issues.

That said, if we wish for binary-compat with past desktop winforms control libraries / apps the only way to do so is to bring back [Serializable] because the use of BinaryFormatter is directly in user code (code spit from resources designer and winforms designer).

**Comment by ViktorHofer** [comment]
cc @danmosemsft

**Comment by Tanya-Solyanik** [comment]
FYI - @merriemcgaw

**Comment by morganbr** [comment]
We should try to apply the same standards here as to other serializable types. Every serializable type should:
1. Have data showing more than one person has ever serialized it (excluding for remoting). I can share my data set from the initial effort.
2. Be checked for potential serialization attack gadgets. While we don't have hard guarantees that serialization on Core is safe, we don't want to introduce anything dangerous if we can avoid it. Types that don't implement ISerializable or any On[De]Serializ[ed/ing] methods aren't likely to be problems.
3. Area owners (and probably @stephentoub) should confirm that we're ok locking in the implementation of the serializable type.

CC @stephentoub, @GrabYourPitchforks

**Comment by danmosemsft** [comment]
@morganbr did you collect any usage data for Winforms though? I looked at your original lists from last summer and you seem to limit it to types in .NET Core (which makes sense). 

Do you have a list of Winforms types that are serializable? ILSPY is not cooperating for me when I look for use of [Serializable] - also I assume there are more assemblies than System.Windows.Forms.dll.

We should also understand what the designer requires to be serializable - presumably we want the designer to work as it does for .NET Framework, even if they seem little serialized in Nuget.

**Comment by morganbr** [comment]
@danmosemsft, I have data for all framework types that I can share internally shortly (just have to dig it up). I don't have the list of all serializable types. ILSpy probably isn't helping because ```[Serializable]``` is actually special metadata, not an attribute.

**Comment by stephentoub** [comment]
> I have data for all framework types

IIRC, that was data gathered based on static analysis; does your tool successfully handle a case like this, where the use of BinaryFormatter is so far removed from where the resulting object is cast to the target type?

As @ericstj highlighted, the key use case for Windows Forms is in the designer storing various drawing-related types into the .resx from the designer, e.g.
![image](https://user-images.githubusercontent.com/2642209/41326383-bf8ede5c-6e8c-11e8-9d7f-75c1001b6b3e.png)
and then using that resource in the app, e.g.
![image](https://user-images.githubusercontent.com/2642209/41326423-ed61e98c-6e8c-11e8-94d4-38f6e11845d0.png)
That results in code like:
```C#
this.pictureBox1.Image = global::WindowsFormsApp1.Properties.Resources.Dog;
```
ending up in the form, which in turn uses `ResourceManager.GetObject`:
```C#
internal static System.Drawing.Bitmap Dog {
    get {
        object obj = ResourceManager.GetObject("Dog", resourceCulture);
        return ((System.Drawing.Bitmap)(obj));
    }
}
```
which in turn indirectly uses `BinaryFormatter` in a deep hierarchy of calls.

As such, I expect we'll find it's very common in Windows Forms apps to use BinaryFormatter with many things you can store as a resource:
![image](https://user-images.githubusercontent.com/2642209/41326757-8fd526d8-6e8e-11e8-9f9a-befe4f2b0a73.png)

**Comment by danmosemsft** [comment]
Discussing with @merriemcgaw what is needed here.

**Comment by danmosemsft** [comment]
Seems clear enough that we need at least Bitmap, Image, Metafile, Icon, and we can get started on those.

**Comment by danmosemsft** [comment]
There is one other big one that gets significant usage -- System.Drawing.Font. Fortunately it is just string, float and two enums.

These enums need `[Serializable]` but do need `[System.Runtime.CompilerServices.TypeForwardedFrom("System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a")]` and a test added - so they're easy:
- System.Drawing.Color
- System.Drawing.ContentAlignment
- System.Drawing.Drawing2D.LinearGradientMode (low usage, but it's trival to add the attribute)
- System.Drawing.FontStyle and GraphicsUnit

**Comment by danmosemsft** [comment]
I have broken out the Enums above into #31477 as they are easy to deal with separately.

**Comment by safern** [comment]
Closed in:  #30817
